### PR TITLE
Normalize ZeRO-3 DeepCompile grad dtype before reduction

### PR DESCRIPTION
**Pull request by tohtana**
Some backward kernels produce gradients in their computation dtype, not necessarily in the parameter storage dtype. For example, if a backward path accumulates or promotes math in fp32, a parameter stored as bf16 can still receive an fp32 raw gradient from that backward computation.

In normal PyTorch execution, that raw gradient reaches the leaf-gradient accumulation step, which stores it according to the tensor's expected grad dtype. ZeRO-3 DeepCompile intercepts the raw compiled-backward gradient before that leaf accumulation boundary. The reducer was assuming the raw gradient dtype was already the expected leaf grad dtype, so it could select an fp32 communication bucket even when the ZeRO grad partition storage was bf16.

To address this, this PR changes `dc.reduce_grad`'s behavior to match PyTorch's leaf-gradient dtype contract. ZeRO-3 registration now records the expected grad dtype for each parameter, and `reduce_grad` normalizes raw compiled-backward gradients to that dtype before selecting the communication bucket.

This follows the documented `grad_dtype` behavior, including preserving explicit `grad_dtype=None` opt-outs: https://docs.pytorch.org/docs/main/generated/torch.sparse.semi_structured.SparseSemiStructuredTensorCUSPARSELT.html#torch.sparse.semi_structured.SparseSemiStructuredTensorCUSPARSELT.grad_dtype

## Testing
- `python -m py_compile deepspeed/compile/init_z3.py tests/unit/compile/test_zero3_grad_dtype.py`
- `python -m pytest -q tests/unit/compile/test_zero3_grad_dtype.py`
I confirmed both passed.

The [full repro](https://gist.github.com/tohtana/9dcf981f381e107073cea71c2948e8ee) would also be useful.

### CODE REVIEW

File: csrc/compile/z3.cpp [hunk 1]
```diff
@@ -268,37 +270,49 @@ class Z3CustomOpExecutor : public CustomOpExecutor {
         for (const ReduceTask& t : reduce_tasks_.at(scalar_type)) {
             auto recv_buf = param_registry_->getParam(t.getDSId()).getGradBuffer();
             bool acc_grad = has_acc_grad_.at(t.getDSId());
+            bool use_tmp_recv = acc_grad || recv_buf.scalar_type() != scalar_type;
```
[the remark this answers is not in the document]
Comment (tohtana) [reply csrc/compile/z3.cpp]: Addressed

**Review by eternalNight — APPROVED** [review]
Looks good to me. Thanks!

**Comment by tohtana** [comment]
@eternalNight Thank you for your review!

**Review by PKUWZP — APPROVED** [review]
Thanks for submitting the PR! If you have any unit test and integration test results available, would be great to attach them as part of the PR description, thanks!

**Comment by tohtana** [comment]
@PKUWZP Thank you for the suggestion. I added testing commands and a link to the [repro](https://gist.github.com/tohtana/9dcf981f381e107073cea71c2948e8ee).
